libraries/app/database_api.cpp

```diff
@@ -1428,38 +1430,47 @@ order_book database_api_impl::get_order_book( const string& base, const string&
    return result;
 }
 
-vector<market_volume> database_api::get_top_markets(uint32_t limit)const
+vector<market_ticker> database_api::get_top_markets(uint32_t limit)const
 {
    return my->get_top_markets(limit);
 }
 
-vector<market_volume> database_api_impl::get_top_markets(uint32_t limit)const
+vector<market_ticker> database_api_impl::get_top_markets(uint32_t limit)const
 {
    FC_ASSERT( _app_options && _app_options->has_market_history_plugin, "Market history plugin is not enabled." );
 
    FC_ASSERT( limit <= 100 );
 
    const auto& volume_idx = _db.get_index_type<graphene::market_history::market_ticker_index>().indices().get<by_volume>();
    auto itr = volume_idx.rbegin();
-   vector<market_volume> result;
+   vector<market_ticker> result;
    result.reserve(limit);
 
-   const fc::time_point_sec now = fc::time_point::now();
-
    while( itr != volume_idx.rend() && result.size() < limit)
    {
-      market_volume mv;
-      mv.time = now;
-      const auto assets = get_assets( { itr->base, itr->quote } );
-      mv.base = assets[0]->symbol;
-      mv.quote = assets[1]->symbol;
-      mv.base_volume = uint128_amount_to_string( itr->base_volume, assets[0]->precision );
-      mv.quote_volume = uint128_amount_to_string( itr->quote_volume, assets[1]->precision );
-      result.emplace_back( std::move(mv) );
+      auto ticker = build_ticker(*itr);
+      result.emplace_back( append_orderbook_to_ticker(ticker) );
       ++itr;
    }
    return result;
 }
+market_ticker database_api_impl::build_ticker(const market_ticker_object& mto) const
+{
+   const auto& assets = get_assets( { mto.base, mto.quote } );
+   const fc::time_point_sec now = _db.head_block_time();
+   market_ticker ticker(mto, now, *assets[0], *assets[1]);
+   return ticker;
+}
```

Comment:
This could be part of the constructor's job. You may want to pass ``now`` and ``assets`` to the constructor to avoid needing references to database_api_impl and _db.